### PR TITLE
Handle some array corner cases in Table constructor.

### PR DESCRIPTION
**Pull request by mdboom**
@taldcroft: While investigating #3634, I discovered some corner cases when passing arrays to the `Table` constructor that I'm surprised didn't work.

- Passing an empty non-recarray array (`np.array([])`) should be treated the same as `None`.

- Passing a 1-D non-recarray could be treated as a single row.  (That's perhaps controversial -- I'd be just as happy raising an exception rather than failing in a confusing way).

- Passing a scalar should raise a nicer exception.

Any thoughts?

### CODE REVIEW

**Comment by taldcroft** [comment]
@mdboom - I agree with what you've done here, thanks.

On the controversial second bullet point, that did give me some pause.  I think it's a reasonable way to go though given that:

- An MxN 2-d array initializes a table as intuitively expected, with M rows.
- A 1-d array in numpy behaves more like a row than column in the sense that `np.vstack` on two 1-d arrays does stack them as if they were horizontal rows.  I.e. for two N element arrays, vstacking ends up with a shape of (2, N).  So from that perspective a 1-d array is a row with unspecified column names.

File: astropy/table/table.py [hunk 1]
```diff
@@ -247,6 +247,11 @@ def __init__(self, data=None, masked=None, names=None, dtype=None,
 
         default_names = None
 
+        if (isinstance(data, np.ndarray) and
+            data.shape == (0,) and
```
Comment (taldcroft) [inline astropy/table/table.py]: Formatting nit to indent the 2nd and 3rd lines here to have a visual offset from the statement block.

Comment (mdboom) [inline astropy/table/table.py]: Fixed.

**Comment by taldcroft** [comment]
:+1: for merge, just needs CHANGES entry.

File: astropy/table/table.py [hunk 2]
```diff
@@ -264,7 +269,13 @@ def __init__(self, data=None, masked=None, names=None, dtype=None,
                 default_names = data.dtype.names
             else:
                 init_func = self._init_from_ndarray  # _homog
-                n_cols = data.shape[1]
+                if data.shape == ():
+                    raise ValueError('Can not initialize a Table with a scalar')
+                elif len(data.shape) == 1:
+                    n_cols = data.shape[0]
+                    data = np.expand_dims(data, axis=0)
```
Comment (mhvk) [inline astropy/table/table.py]: Totally trivial, and only if you haven't done the last change yet, but writing this as
```
data = data[np.newaxis, :]
```
or (bit less good)
```
data = data.reshape(1, -1)
```
is clearer to me at least.

(Actually, if you do this before the `n_cols` assignment, you can remove the `else:` clause.)

**Comment by mdboom** [comment]
Comments are addressed.

**Comment by taldcroft** [comment]
@mdboom - sorry, on my final lookthrough I realized it would be good to add a documentation example.  I've put in a PR on your branch.

**Comment by taldcroft** [comment]
Windows failing because of int32 default, grrr..

I think the doc example can be `Table(np.array([1.0, 2.0, 3.0]), names=['a', 'b', 'c']) ` which will be `float64` on linux and windows.

**Pull request merged**